src/Nethermind/Nethermind.Blockchain/TxPools/TxPool.cs

```diff
@@ -244,52 +230,76 @@ public AddTxResult AddTransaction(Transaction tx, long blockNumber, bool isOwn =
              * if we leave it for block production only.
              * */
             
-            if (isOwn && !HandleOwnTransaction(tx))
+            if (CheckOwnTransactionAlreadyUsed(tx, isOwn))
             {
-                _transactions.TryRemove(tx.Hash, out _);
                 return AddTxResult.OwnNonceAlreadyUsed;
             }
-            
+
+            if (!_transactions.TryInsert(tx.Hash, tx))
+            {
+                // If transaction is fresh and already known then it may be stored in memory.
+                Metrics.PendingTransactionsKnown++;
+                return AddTxResult.AlreadyKnown;
+            }
+
+            if (_txStorage.Get(tx.Hash) != null)
+            {
+                // If transaction is a bit older and already known then it may be stored in the persistent storage.
+                Metrics.PendingTransactionsKnown++;
+                return AddTxResult.AlreadyKnown;
+            }
+
+            HandleOwnTransaction(tx, isOwn);
+
             NotifySelectedPeers(tx);
             FilterAndStoreTx(tx, blockNumber);
             NewPending?.Invoke(this, new TxEventArgs(tx));
             return AddTxResult.Added;
         }
 
-        private bool HandleOwnTransaction(Transaction transaction)
+        private void HandleOwnTransaction(Transaction transaction, bool isOwn)
         {
-            var address = transaction.SenderAddress;
-            lock (_locker)
+            if (isOwn)
             {
-                if (!_nonces.TryGetValue(address, out var addressNonces))
-                {
-                    var currentNonce = _stateProvider.GetNonce(address);
-                    addressNonces = new AddressNonces(currentNonce);
-                    _nonces.TryAdd(address, addressNonces);
-                }
-                
-                if (!addressNonces.Nonces.TryGetValue(transaction.Nonce, out var nonce))
-                {
-                    nonce = new Nonce(transaction.Nonce);
-                    addressNonces.Nonces.TryAdd(transaction.Nonce, new Nonce(transaction.Nonce));
-                }
+                _ownTransactions.TryAdd(transaction.Hash, transaction);
+                _ownTimer.Enabled = true;
+                if (_logger.IsInfo) _logger.Info($"Broadcasting own transaction {transaction.Hash} to {_peers.Count} peers");
+            }
+        }
 
-                if (!(nonce.TransactionHash is null && nonce.TransactionHash != transaction.Hash))
+        private bool CheckOwnTransactionAlreadyUsed(Transaction transaction, bool isOwn)
+        {
+            if (isOwn)
```

Comment:
why was locking removed?

Reply:
i wasn't, its still right here in line 275